compiler/rustc_hir_analysis/src/check/check.rs

```diff
@@ -465,7 +468,51 @@ fn check_opaque_meets_bounds<'tcx>(
         }
     }
     // Clean up after ourselves
-    let _ = infcx.take_opaque_types();
+    for (key, ty) in infcx.take_opaque_types() {
+        let found_hidden_ty = infcx.resolve_vars_if_possible(ty.hidden_type.ty);
+        if found_hidden_ty.is_ty_var() {
+            // Nothing was actually constrained.
+            continue;
+        }
+        if let ty::Alias(ty::Opaque, alias) = found_hidden_ty.kind() {
+            if alias.def_id == key.def_id.to_def_id() {
+                if alias.substs != key.substs {
+                    // FIXME(oli-obk): figure out why tests/ui/impl-trait/impl-fn-predefined-lifetimes.rs
+                    // hits this code path.
+                    tcx.sess.delay_span_bug(
+                        tcx.def_span(alias.def_id),
+                        format!("{:?} != {:?}", alias.substs, key.substs),
+                    );
+                }
+                // Nothing was actually constrained, this is an opaque usage that was
+                // only discovered to be opaque after inference vars resolved.
+                continue;
+            }
+        }
+        let found_hidden_ty = found_hidden_ty.fold_with(&mut BottomUpFolder {
+            tcx,
+            ty_op: |t| t,
+            ct_op: |c| c,
+            lt_op: |l| match l.kind() {
+                RegionKind::ReVar(_) => tcx.lifetimes.re_erased,
+                _ => l,
+            },
+        });
+        let hidden_ty = tcx.type_of(key.def_id).subst(tcx, key.substs);
+        if hidden_ty != found_hidden_ty {
```

Comment:
wwwill be fixed by making `InternalSubsts::identity_for_item(tcx, def_id.to_def_id());` use the anchor def id, instead of the opaque def id. For RPIT we make the hidden type based on the function, not the RPIT generics (which are just copies of the function generics, but with different DefIds)